relayer/app/cursor/cursors.go

```diff
@@ -0,0 +1,156 @@
+package cursor
+
+import (
+	"context"
+	"slices"
+	"time"
+
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/expbackoff"
+	"github.com/omni-network/omni/lib/log"
+	"github.com/omni-network/omni/lib/netconf"
+	"github.com/omni-network/omni/lib/xchain"
+
+	db "github.com/cosmos/cosmos-db"
+)
+
+// Cursors implements operations on the persisted cursors for network.
+type Cursors struct {
+	db              CursorTable
+	confirmInterval time.Duration
+	xProvider       xchain.Provider
+	streams         []stream
+	network         netconf.Network
+}
+
+func NewCursors(
+	db db.DB,
+	xProvider xchain.Provider,
+	confirmInterval time.Duration,
+	network netconf.Network,
+) (*Cursors, error) {
+	cursorsTable, err := NewCursorsTable(db)
+	if err != nil {
+		return nil, err
+	}
+
+	var streams []stream
+	for _, chain := range network.EVMChains() {
+		for _, streamID := range network.StreamsTo(chain.ID) {
+			s := stream{
+				SrcVersion: xchain.ChainVersion{
+					ID:        streamID.SourceChainID,
+					ConfLevel: streamID.ConfLevel(),
+				},
+				DestChain: chain,
+				Network:   network,
+			}
+			streams = append(streams, s)
+		}
+	}
+
+	return &Cursors{
+		db:              cursorsTable,
+		confirmInterval: confirmInterval,
+		xProvider:       xProvider,
+		streams:         streams,
+		network:         network,
+	}, nil
+}
+
+// ConfirmedOffset loads confirmed cursor offset for the chain version and destination chain ID.
+// If a cursor doesn't exist a 0, false is returned.
+func (c *Cursors) ConfirmedOffset(
+	ctx context.Context,
+	srcVersion xchain.ChainVersion,
+	destChain netconf.Chain,
+) (uint64, bool, error) {
+	cursors, err := getAllStreamCursors(ctx, c.db, stream{srcVersion, destChain, c.network})
+	if err != nil {
+		return 0, false, err
+	}
+
+	slices.Reverse(cursors)
+	for _, c := range cursors {
+		if c.GetConfirmed() {
+			return c.GetAttestOffset(), true, nil
+		}
+	}
+
+	return 0, false, nil
+}
+
+// Save a cursor to the storage.
+// If a cursor already exists it is updated with the provided data.
+func (c *Cursors) Save(
+	ctx context.Context,
+	srcVersion xchain.ChainVersion,
+	destChain netconf.Chain,
+	attestationOffset uint64,
+	empty bool,
+) error {
+	cursor := &Cursor{
+		SrcChainId:   srcVersion.ID,
+		ConfLevel:    uint32(srcVersion.ConfLevel),
+		DstChainId:   destChain.ID,
+		AttestOffset: attestationOffset,
+		Confirmed:    false,
+		Empty:        empty,
+	}
+
+	err := c.db.Save(ctx, cursor)
+	if err != nil {
+		return errors.Wrap(err, "insert cursor")
+	}
+
+	log.Debug(ctx, "New cursor saved",
+		"src_chain", srcVersion.ID,
```

Comment:
never log chain ID, always log human readable strings. I suggest you use `c.network.ChainVersionName(srcVersion` and `destChain.Name`.